### PR TITLE
expression: implement vectorized evaluation for ‘builtinFloorDecToDecSig’

### PR DESCRIPTION
**Pull request by Zhuchengyu04**
PCP  #12102

### What problem does this PR solve? 
implement vectorized evaluation for   builtinFloorDecToDecSig

### What is changed and how it works?
```
goos: windows
goarch: amd64
pkg: github.com/pingcap/tidb/expression
BenchmarkVectorizedBuiltinMathFunc/builtinFloorDecToDecSig-VecBuiltinFunc-8                 6014            255691 ns/op           38544 B/op        803 allocs/op
BenchmarkVectorizedBuiltinMathFunc/builtinFloorDecToDecSig-NonVecBuiltinFunc-8              4005            267289 ns/op           57072 B/op       1189 allocs/op
PASS
ok      github.com/pingcap/tidb/expression      2.940s
```

### Check List 

Tests 

 - Unit tests

### CODE REVIEW

**Comment by Zhuchengyu04** [comment]
> #13314 has been dealing with `builtinCeilDecToDecSig`, could you please fix the `builtinFloorDecToDecSig` in this PR?

With pleasure

File: expression/builtin_math_vec.go [hunk 1]
```diff
@@ -853,15 +853,54 @@ func (b *builtinCeilDecToDecSig) vectorized() bool {
 }
 
 func (b *builtinCeilDecToDecSig) vecEvalDecimal(input *chunk.Chunk, result *chunk.Column) error {
-	return errors.Errorf("not implemented")
+		return errors.Errorf("not implemented")
```
Comment (francis0407) [inline expression/builtin_math_vec.go]: Please revert this change.

File: expression/builtin_math_vec.go [hunk 2]
```diff
@@ -853,15 +853,54 @@ func (b *builtinCeilDecToDecSig) vectorized() bool {
...
+	result.MergeNulls(buf)
+
+	res := result.Decimals()
+	bufs := buf.Decimals()
+
+	for i := 0; i < n; i++ {
+		if result.IsNull(i) {
+			continue
+		}
+		result := new(types.MyDecimal)
```
Comment (francis0407) [inline expression/builtin_math_vec.go]: We'd better not use the name `result` here, it's really confusing.

File: expression/builtin_math_vec.go [hunk 3]
```diff
@@ -853,15 +853,54 @@ func (b *builtinCeilDecToDecSig) vectorized() bool {
...
+		if bufs[i].IsNegative() {
+			err = res[i].Round(result, 0, types.ModeTruncate)
```
Comment (francis0407) [inline expression/builtin_math_vec.go]: `res[i].Round(...)` seems not right. You should use `bufs[i].Round(...)` here?

File: expression/builtin_math_vec_test.go [hunk 4]
```diff
@@ -95,6 +95,7 @@ var vecBuiltinMathCases = map[string][]vecExprBenchCase{
 		{retEvalType: types.ETInt, childrenTypes: []types.EvalType{types.ETDecimal}, geners: nil},
 		{retEvalType: types.ETDecimal, childrenTypes: []types.EvalType{types.ETInt}, geners: nil},
 		{retEvalType: types.ETDecimal, childrenTypes: []types.EvalType{types.ETInt}, childrenFieldTypes: []*types.FieldType{{Tp: mysql.TypeLonglong, Flag: mysql.UnsignedFlag}}, geners: nil},
+		{retEvalType: types.ETDecimal, childrenTypes: []types.EvalType{types.ETDecimal}, geners: nil},
```
Comment (francis0407) [inline expression/builtin_math_vec_test.go]: I have checked the `getFunction` method of the `Floor` expression. You need to make sure the childrenType (Decimal) statifies the conditoin: `fieldTp.Flen-fieldTp.Decimal > mysql.MaxIntWidth-2`. 
So I suggest you to change this test case to:
```
{retEvalType: types.ETDecimal, childrenTypes: []types.EvalType{types.ETDecimal},
  childrenFieldTypes: []*types.FieldType{{Tp: mysql.TypeNewDecimal, Flen: 32, Decimal: 2}}, geners: nil},
```

Comment (Zhuchengyu04) [reply expression/builtin_math_vec_test.go]: thank you

File: expression/builtin_math_vec_test.go [hunk 5]
```diff
@@ -95,6 +95,7 @@ var vecBuiltinMathCases = map[string][]vecExprBenchCase{
 		{retEvalType: types.ETInt, childrenTypes: []types.EvalType{types.ETDecimal}, geners: nil},
 		{retEvalType: types.ETDecimal, childrenTypes: []types.EvalType{types.ETInt}, geners: nil},
 		{retEvalType: types.ETDecimal, childrenTypes: []types.EvalType{types.ETInt}, childrenFieldTypes: []*types.FieldType{{Tp: mysql.TypeLonglong, Flag: mysql.UnsignedFlag}}, geners: nil},
+		{retEvalType: types.ETDecimal, childrenTypes: []types.EvalType{types.ETDecimal}, childrenFieldTypes: []*types.FieldType{{Tp: mysql.TypeNewDecimal, Flen: 32, Decimal: 2}}, geners: nil}},
```
Comment (francis0407) [inline expression/builtin_math_vec_test.go]: There is a redundent '}' at the end.

Comment (Zhuchengyu04) [reply expression/builtin_math_vec_test.go]: ok,i have amended it

File: expression/builtin_math_vec.go [hunk 6]
```diff
@@ -857,11 +857,50 @@ func (b *builtinCeilDecToDecSig) vecEvalDecimal(input *chunk.Chunk, result *chun
...
 
 func (b *builtinFloorDecToDecSig) vectorized() bool {
-	return false
+	return true
 }
 
 func (b *builtinFloorDecToDecSig) vecEvalDecimal(input *chunk.Chunk, result *chunk.Column) error {
-	return errors.Errorf("not implemented")
+	n := input.NumRows()
+	buf, err := b.bufAllocator.get(types.ETDecimal, n)
```
Comment (francis0407) [inline expression/builtin_math_vec.go]: I think you don't need to allocate a new `buf` for this function. Both `buf` and `res` are Decimals, you can just reuse the `res` here.

Comment (Zhuchengyu04) [reply expression/builtin_math_vec.go]: sorry, the type of buf is Column, but type of res is Decimal

Comment (francis0407) [reply expression/builtin_math_vec.go]: `buf` will be converted to `bufs` which is a slice of Decimal.
I mean you can use the `input` to replace the `buf`.

Comment (Zhuchengyu04) [reply expression/builtin_math_vec.go]: OK，i'll amend it

File: expression/builtin_math_vec.go [hunk 7]
```diff
@@ -857,11 +857,38 @@ func (b *builtinCeilDecToDecSig) vecEvalDecimal(input *chunk.Chunk, result *chun
...
+	if err := b.args[0].VecEvalDecimal(b.ctx, input, result); err != nil {
+		return err
+	}
+	ds := result.Decimals()
+
+	for i := 0; i < n; i++ {
+		if result.IsNull(i) {
+			continue
+		}
+		rst := new(types.MyDecimal)
```
Comment (qw4990) [inline expression/builtin_math_vec.go]: Please define this variable outside this loop so that we don't need to create a new object each time.

Comment (qw4990) [reply expression/builtin_math_vec.go]: Maybe we can optimize it later.
